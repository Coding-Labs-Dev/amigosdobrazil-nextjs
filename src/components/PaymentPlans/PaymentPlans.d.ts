import { PaymentPlan, TransportPlan } from '~/types/models';

export interface Props {
  slug: string;
  transportPlans: TransportPlan[];
  paymentPlans: PaymentPlan[];
  activePlanIndex: number;
  display?: 'xs' | 'md';
}
