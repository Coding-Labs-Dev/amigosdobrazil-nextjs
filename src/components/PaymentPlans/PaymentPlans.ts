export interface PaymentPlan {
  id: string;
  date: Date;
  usd: string;
  brl: string;
  downPayment: string;
  installmentsQty: number;
  installmentsValue: string;
}

export interface Props {
  slug: string;
  paymentPlans: PaymentPlan[];
  display?: 'xs' | 'md';
}
