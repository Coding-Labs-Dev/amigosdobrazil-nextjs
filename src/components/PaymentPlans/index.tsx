import React from 'react';
import moment from 'moment';

import Link from 'next/link';
import {
  Container,
  Card,
  CardTitle,
  CardBody,
  Row,
  Col,
  Button,
} from 'reactstrap';
import { FaCheckCircle, FaTimesCircle } from 'react-icons/fa';

import { Props } from './PaymentPlans';
import { Wrapper } from './styles';

const PaymentPlans: React.FC<Props> = ({
  slug,
  paymentPlans,
  transportPlans,
  display = 'xs',
  activePlanIndex,
}) => {
  return (
    <Wrapper className="my-3">
      <Container>
        <Card
          className={`pt-5 pb-3 mb-5 px-4 px-md-5 shadow border-0 ${
            display === 'xs' ? 'd-block d-md-none' : 'd-none d-md-block'
          }`}
        >
          <CardTitle>
            <h1 className="paymentPlan-title">Planos de Pagamento</h1>
          </CardTitle>
          <CardBody>
            {!!transportPlans.length && (
              <h1 className="part-title">Parte Aérea</h1>
            )}
            {transportPlans.map(plan => (
              <Card
                className="p-4 my-3 b-payment-plan border-primary b-payment-plan__active"
                key={`${display}_tansport_plan_${plan.id}`}
                style={{ borderWidth: 1 }}
              >
                <Row>
                  <Col xs={12} lg={8} className="align-items-stretch">
                    <Row className="text-left h-100">
                      <Col xs={1} className="my-auto pr-4">
                        <FaCheckCircle size={20} className="text-primary" />
                      </Col>
                      <Col className="my-auto">
                        <span className="font-weight-bold text-primary d-block">
                          Classe econômica
                        </span>
                        <span className="font-weight-bold text-light d-block">
                          {Number(plan.usd).toLocaleString('pt', {
                            style: 'currency',
                            currency: 'USD',
                            currencyDisplay: 'symbol',
                          })}
                        </span>
                      </Col>
                    </Row>
                  </Col>
                  <Col xs={12} lg={4} className="text-right my-auto">
                    <span className="text-light small d-block">
                      + Taxa de Embarque
                    </span>
                    <span className="text-light small d-block">
                      Em até {plan.installmentsQty}x sem juros
                    </span>
                  </Col>
                </Row>
              </Card>
            ))}
            <h1 className={`part-title ${!!transportPlans.length && 'pt-4'}`}>
              Parte Terrestre
            </h1>
            {paymentPlans.map((plan, i) => {
              const active = i === activePlanIndex;
              return (
                <Card
                  className={`p-4 my-3 b-payment-plan ${
                    active
                      ? 'border-primary b-payment-plan__active'
                      : 'border-light'
                  }`}
                  key={`${display}_${plan.id}`}
                  style={{ borderWidth: 1, opacity: active ? 1 : 0.6 }}
                >
                  <Row>
                    <Col xs={12} lg={8} className="align-items-stretch">
                      <Row className="text-left h-100">
                        <Col xs={1} className="my-auto pr-4">
                          {active ? (
                            <FaCheckCircle size={20} className="text-primary" />
                          ) : (
                            <FaTimesCircle
                              size={20}
                              className="text-danger"
                              style={{ opacity: 0.5 }}
                            />
                          )}
                        </Col>
                        <Col className="my-auto">
                          <span
                            className={`font-weight-bold text-${
                              active ? 'primary' : 'light'
                            } d-block`}
                          >
                            Inscrição
                            {i < paymentPlans.length - 1 ? ' até ' : ' após '}
                            {moment(plan.date).format('DD/MM/YYYY')}
                          </span>
                          <span className="font-weight-bold text-light d-block">
                            {Number(plan.usd).toLocaleString('pt', {
                              style: 'currency',
                              currency: 'USD',
                              currencyDisplay: 'symbol',
                            })}
                          </span>
                        </Col>
                      </Row>
                    </Col>
                    <Col xs={12} lg={4} className="text-right my-auto">
                      {Number(plan.downPayment) ? (
                        <>
                          <span className="text-light small d-block">
                            Entrada
                          </span>
                          <span
                            className={`font-weight-bold text-${
                              active ? 'dark' : 'light'
                            } d-block`}
                          >
                            {Number(plan.downPayment).toLocaleString('pt', {
                              style: 'currency',
                              currency: 'BRL',
                            })}
                          </span>
                        </>
                      ) : (
                        <>
                          <span className="text-light small d-block">
                            Taxa de Inscrição
                          </span>
                        </>
                      )}
                      <span className="text-light small d-block">
                        + {plan.installmentsQty}x{' '}
                        {Number(plan.installmentsValue).toLocaleString('pt', {
                          style: 'currency',
                          currency: 'BRL',
                        })}
                      </span>
                    </Col>
                  </Row>
                </Card>
              );
            })}
            <Row className="px-3">
              {activePlanIndex >= 0 ? (
                <Link href={`/roteiros/${slug}/reservar`}>
                  <Button
                    color="warning"
                    className="ml-auto text-uppercase mt-2"
                  >
                    Reservar
                  </Button>
                </Link>
              ) : (
                <Button
                  color="warning"
                  className="ml-auto text-uppercase mt-2"
                  disabled
                >
                  Reservar
                </Button>
              )}
            </Row>
          </CardBody>
        </Card>
      </Container>
    </Wrapper>
  );
};

export default PaymentPlans;
