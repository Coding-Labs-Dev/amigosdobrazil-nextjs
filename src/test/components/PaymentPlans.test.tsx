import React from 'react';
import renderer from 'react-test-renderer';
import MockDate from 'mockdate';
import { shallow } from 'enzyme';

import PaymentPlans from '~/components/PaymentPlans';
import { Props } from '~/components/PaymentPlans/PaymentPlans';

describe('PaymentPlans', () => {
  beforeEach(() => {
    MockDate.set('2020-06-22');
  });

  afterEach(() => {
    MockDate.reset();
  });

  const setupComponent = (props: Partial<Props> = {}) => (
    <PaymentPlans
      slug="india-01-2021"
      paymentPlans={[
        {
          id: '1',
          date: new Date('2020-05-30T00:00:00+0000'),
          usd: '4000.00',
          brl: '20000.00',
          downPayment: '0.00',
          installmentsQty: 10,
          installmentsValue: '2000.00',
        },
      ]}
      {...props}
    />
  );

  it('renders', () => {
    const component = renderer.create(setupComponent());
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('enables the first payment plan when it\'s after today\'s date', () => {
    const paymentPlans = [
      {
        id: '1',
        date: new Date('2020-06-30T00:00:00+0000'),
        usd: '4000.00',
        brl: '20000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '2000.00',
      },
      {
        id: '2',
        date: new Date('2020-07-30T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
      {
        id: '3',
        date: new Date('2020-07-31T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
    ];

    const component = shallow(setupComponent({ paymentPlans }));
    expect(component.find('.b-payment-plan').at(0)).toHaveClassName('.b-payment-plan__active');

  });

  it('enables the next payment plan after today\'s date', () => {
    const paymentPlans = [
      {
        id: '1',
        date: new Date('2020-05-30T00:00:00+0000'),
        usd: '4000.00',
        brl: '20000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '2000.00',
      },
      {
        id: '2',
        date: new Date('2020-07-30T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
      {
        id: '3',
        date: new Date('2020-07-31T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
    ];

    const component = shallow(setupComponent({ paymentPlans }));
    expect(component.find('.b-payment-plan').at(1)).toHaveClassName('.b-payment-plan__active');

  });

  it('enables the latest payment plan after today\'s date', () => {
    const paymentPlans = [
      {
        id: '1',
        date: new Date('2020-05-20T00:00:00+0000'),
        usd: '4000.00',
        brl: '20000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '2000.00',
      },
      {
        id: '2',
        date: new Date('2020-06-20T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
      {
        id: '3',
        date: new Date('2020-06-21T00:00:00+0000'),
        usd: '8000.00',
        brl: '40000.00',
        downPayment: '0.00',
        installmentsQty: 10,
        installmentsValue: '4000.00',
      },
    ];

    const component = shallow(setupComponent({ paymentPlans }));
    expect(component.find('.b-payment-plan').at(2)).toHaveClassName('.b-payment-plan__active');

  });
});
