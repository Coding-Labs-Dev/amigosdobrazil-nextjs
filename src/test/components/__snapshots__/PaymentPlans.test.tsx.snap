// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentPlans renders 1`] = `
<div
  className="sc-AxjAm StDqN my-3"
>
  <div
    className="container"
  >
    <div
      className="pt-5 pb-3 mb-5 px-4 px-md-5 shadow border-0 d-block d-md-none card"
    >
      <div
        className="card-title"
      >
        <h1
          className="paymentPlan-title"
        >
          Planos de Pagamento
        </h1>
      </div>
      <div
        className="card-body"
      >
        <div
          className="p-4 my-3 b-payment-plan border-primary b-payment-plan__active card"
          style={
            Object {
              "borderWidth": 1,
              "opacity": 1,
            }
          }
        >
          <div
            className="row"
          >
            <div
              className="align-items-stretch col-12 col-lg-8"
            >
              <div
                className="text-left h-100 row"
              >
                <div
                  className="my-auto pr-4 col-1"
                >
                  <svg
                    className="text-primary"
                    fill="currentColor"
                    height={20}
                    size={20}
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 512 512"
                    width={20}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                    />
                  </svg>
                </div>
                <div
                  className="my-auto col"
                >
                  <span
                    className="font-weight-bold text-primary d-block"
                  >
                    Inscrição até
                     
                    29/05/2020
                  </span>
                  <span
                    className="font-weight-bold text-light d-block"
                  >
                    US$ 4.000,00
                  </span>
                </div>
              </div>
            </div>
            <div
              className="text-right my-auto col-12 col-lg-4"
            >
              <span
                className="text-light small d-block"
              >
                Entrada
              </span>
              <span
                className="font-weight-bold text-dark d-block"
              >
                R$ 0,00
              </span>
              <span
                className="text-light small d-block"
              >
                + 
                10
                x
                 
                R$ 2.000,00
              </span>
            </div>
          </div>
        </div>
        <div
          className="px-3 row"
        >
          <button
            aria-label={null}
            className="ml-auto text-uppercase mt-2 btn btn-warning"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            Reservar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
